compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -5690,7 +5748,11 @@ private BType getErrorType(BType collectionType, BLangQueryExpr queryExpr, Analy
         return errorType;
     }
 
-    private BType getNonContextualQueryType(BType staticType, BType basicType) {
+    private BType getNonContextualQueryType(BLangQueryExpr queryExpr, BType staticType, BType basicType) {
+        if (queryExpr.isMap) {
+            return symTable.mapType;
+        }
```

Comment:
I think, here we cannot decide the `resultType` by checking the `basicType`. We are passing collectionType for the basicType argument.

For an example, 
```ballerina
var a = map from string|int a in ["a", "b", 1, 2]
                let [string, int]|[string, string] c = a
                select c;
```
Here, basicType is `[string,string,int,int]`.

So I think, here, we have to use the `isMap` flag, because there is no other way to identify whether the queryExpr is used to construct a map or not. WDYT @pcnfernando

Reply:
I am a bit confused about this logic. Please consider following cases.
case 1
```ballerina
    map<int|string> _ = map from var user in users
                 where user.age > 21 && user.age < 60
                 select user;
```
case 2
```ballerina
    var _ = map from var user in users
                 where user.age > 21 && user.age < 60
                 select user;
```
For case 1, resolvedType is `other` but for case 2, resolvedType is `map`, but by looking at the select expression, resolvedType should be other for both cases. 
I think resolvedType should not depend on the type of lhs.

Reply:
If the CEP type is not there and the constructor type is there; we should consider the constructor type as the basic type. Shall we check the same being followed in other places as well when CEP is not there? @SandaruJayawardana 
>A query-expr that constructs a mapping must start with the map keyword. An emitted value [k, v] results in adding a field with key k and value v.
If the query-expr starts with table, then the query-expr will construct a table; the key-specifier specifies the key sequence of the constructed table in the same way as with a table-constructor-expr. **_Otherwise_**, the applicable contextually expected type determines the basic type of the value constructed. If there is no contextually expected type, then the basic type of the value constructed is the basic type of expression following in; it is a compile-time error if the static type of this expression is not a subtype of one of the basic types that a query expression can construct.

Reply:
`A query-expr that constructs a mapping must start with the map keyword` according to this sentence, `map` keyword should be there to construct a map with query expr. So I think we don't check the collection type for the map case.

Reply:
Better set map as the type with map constructors and for others to call this method